torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu

```diff
@@ -269,15 +278,30 @@ class NCCLSymmetricMemoryAllocator : public SymmetricMemoryAllocator {
   };
 
   c10::intrusive_ptr<SymmetricMemory> rendezvous(
-      void* ptr,  // data_ptr() of the tensor
+      const at::Tensor& tensor,
       const std::optional<std::string>& group_name) override {
     TORCH_CHECK(group_name.has_value(), "group_name must be provided");
+
+    // Using raw address of TensorImpl as a unique key of tensor in `symm_mems_`
+    // map, because other addresses such as `tensor.data_ptr()` or
+    // `tensor.storage().data_ptr()` may been shared by views and slices.
+    auto tensor_raw_ptr = (void*)tensor.unsafeGetTensorImpl();
+    auto symm_mem_key = std::make_tuple(tensor_raw_ptr, *group_name);
     {
-      auto it = symm_mems_.find(std::make_tuple(ptr, *group_name));
+      auto it = symm_mems_.find(symm_mem_key);
       if (it != symm_mems_.end()) {
-        return it->second;
+        auto symm_mem = it->second;
+        if (!symm_mem->expired()) {
```

Comment:
if symm_mem is expired (i.e. tensor got deleted) you should destruct it here, to avoid leaking resources?